lib/factories/default-eventbus.js

```diff
@@ -0,0 +1,373 @@
+ *      dotted object path.
+ *      Default: undefined
+ *
+ * - id_field
+ *      This field will be used as the event's 'id' in log messages
+ *      Default: undefined
+ *
+ * - kafka.conf
+ *      node-rdkafka KafkaProducer configuration
+ *
+ * - kafka.topic_conf
+ *      node-rdkafka KafkaProducer topic configuration
+ */
+
+
+const defaultOptions = {
+    schema_uri_field:       '$schema',
+    schema_base_uri:        undefined,
+    schema_file_extension:  undefined,
+    stream_field:           undefined,
+    topic_prefix:           undefined,
+    key_field:              undefined,
+    partition_field:        undefined,
+    id_field:               undefined,
+    kafka: {
+        conf: {
+            'metadata.broker.list': 'localhost:9092'
+        },
+        topic_conf: {}
+    },
+};
+
+/**
+ * Returns a function that extracts the event's schema URI.
+ * @param {Object} options
+ * @param {string} options.schema_uri_field
+ *      Used to extract the event's schema URI.
+ * @return {function(Object): string}
+ */
+function makeExtractSchemaUri(options) {
+    const schemaUriField = _.get(options, 'schema_uri_field', defaultOptions.schema_uri_field);
+    return (event) => {
+        try {
+            return objectGet(event, _.get(options, 'schema_uri_field'));
+        } catch (err) {
+            // Wrap PropertyNotFoundError in a new specific Error about missing Schema URI.
+            if (err instanceof PropertyNotFoundError) {
+                throw new EventSchemaUriMissingError(
+                    'Event schema URI cannot be extracted. ' +
+                    `Event must have a '${schemaUriField}' property`
+                );
+            }
+        }
+    };
+}
+
+/**
+ * Returns a function that given a URI, will resolve it using
+ * options.schema_base_uri and options.schema_file_extension.
+ * See also the event-utils resolveUri function.
+ * @param {Object} options
+ * @param {string} options.schema_base_uri
+ *      If set, this is prefixed to un-anchored schema URIs.
+ * @param {string} options.schema_file_extension
+ *      If set, this is suffixed to schema URIs that dont' already have a file extension.
+ * @return {function(string): string}
+ */
+function makeResolveSchemaUri(options) {
+    return uri => resolveUri(
+        uri, options.schema_base_uri, options.schema_file_extension
+    );
+}
+
+/**
+ * If options.stream_field is configured, then it will be used to extract
+ * the event's destination stream name.  Otherwise, it is assumed
+ * that the event's sanitized schema URI should be used as the stream name.
+ * The schema URI returned by the function returned by makeExtractSchemaUri
+ * with options.schema_uri_field will be used to extract the schema URI,
+ * and then the schema URI will be sanitized to remove potential bad charachters.
+ * Any character that does not match [A-Za-z0-9_-] will be replaced with
```

Comment:
As we've learned, it's better not to mix underscores and dots into kafka names. We will probably prepend with DC name with a dot (in wikimedia-eventbus, but whatever, let's better use `-` everywhere)

Reply:
Ah hm right.  I added '.' into allowed characters, even if folks shouldn't do it, we can let them (and us) use them anyway.